### PR TITLE
MM-23408: Added channel ID check for Plugin API UploadFile

### PR DESCRIPTION
**Pull request by waqasraz**
#### Summary

Added the 2 checks to Plugin API [UploadFile](https://github.com/mattermost/mattermost-server/blob/3e2175f897c46c0d33da0de83db996dc9d97a8f5/app/file.go#L437-L453)
   1. check that the channel ID exists, returning an error if it does not.
   2. check that the user uploading the file has permissions over that channel, returning an error if the user is not authorized

Also added translation string for the errors defined in the UploadFile method.

#### Ticket Link

JIRA: https://mattermost.atlassian.net/browse/MM-23408
Fixes https://github.com/mattermost/mattermost-server/issues/14108

### CODE REVIEW

File: app/file.go [hunk 1]
```diff
@@ -441,6 +441,17 @@ func (a *App) UploadFile(data []byte, channelId string, filename string) (*model
 		return nil, appError
 	}
 
+	_, err := a.GetChannel(channelId)
+	if err != nil {
```
Comment (agarciamontoro) [inline app/file.go]: Nit: could we follow the Go convention here and write this as follows?
```go
if _, err := a.GetChannel(channelId); err != nil {
```

File: app/file_test.go [hunk 2]
```diff
@@ -87,13 +87,34 @@ func TestDoUploadFile(t *testing.T) {
 }
 
 func TestUploadFile(t *testing.T) {
-	th := Setup(t)
+	th := Setup(t).InitBasic()
 	defer th.TearDown()
 
-	channelId := model.NewId()
+	session, _ := th.App.CreateSession(
```
Comment (agarciamontoro) [inline app/file_test.go]: We should require that the error here is `nil`; otherwise, the test could fail and we would not know why.

File: app/file_test.go [hunk 3]
```diff
@@ -87,13 +87,34 @@ func TestDoUploadFile(t *testing.T) {
...
+	channelId := th.BasicChannel.Id
 	filename := "test"
 	data := []byte("abcd")
 
+	info1, _ := th.App.UploadFile(data, model.NewId(), filename)
+	require.Nil(t, info1, "Channel ID does not exist.")
+
+	// remove permission to upload file.
+	th.RemovePermissionFromRole(model.PERMISSION_UPLOAD_FILE.Id, model.CHANNEL_USER_ROLE_ID)
+	info1, _ = th.App.UploadFile(data, channelId, filename)
```
Comment (agarciamontoro) [inline app/file_test.go]: I would also check here that the error is as we expect it is, a non-nil AppError.

File: app/file_test.go [hunk 4]
```diff
@@ -87,13 +87,34 @@ func TestDoUploadFile(t *testing.T) {
...
+	require.Nil(t, info1, "No Permission to upload file for this user.")
+
+	//restore permission
+	th.AddPermissionToRole(model.PERMISSION_UPLOAD_FILE.Id, model.CHANNEL_USER_ROLE_ID)
+
```
Comment (agarciamontoro) [inline app/file_test.go]: Nit: could we remove this newline so both checks, the one with permissions and the one without permissions, have the same structure?

File: app/plugin_api_test.go [hunk 5]
```diff
@@ -1119,6 +1119,16 @@ func TestPluginCreatePostWithUploadedFile(t *testing.T) {
...
 
+	session, _ := th.App.CreateSession(
+		&model.Session{
+			UserId:   th.BasicUser.Id,
+			CreateAt: model.GetMillis(),
+			Roles:    model.CHANNEL_USER_ROLE_ID,
+			IsOAuth:  false,
+		})
+
+	th.App.SetSession(session)
```
Comment (agarciamontoro) [inline app/plugin_api_test.go]: Good catch!

File: i18n/es.json [hunk 6]
```diff
@@ -1192,6 +1192,14 @@
     "id": "api.file.upload_file.incorrect_number_of_files.app_error",
     "translation": "No se pudo subir los archivos. El número de archivos especificado es incorrecto."
   },
+  {
+    "id": "api.file.upload_file.incorrect_channelId.app_error",
+    "translation": "No se pueden cargar archivos. Id. De canal incorrecto: {{.channelId}}"
```
Comment (agarciamontoro) [inline i18n/es.json]: I'd write this as `No se pudo subir el archivo. Identificador de canal incorrecto: {{.channelId}}.`

File: app/file.go [hunk 7]
```diff
@@ -441,6 +441,17 @@ func (a *App) UploadFile(data []byte, channelId string, filename string) (*model
...
+		return nil, model.NewAppError("UploadFiles", "api.file.upload_file.incorrect_channelId.app_error",
+			map[string]interface{}{"channelId": channelId}, "", http.StatusBadRequest)
+	}
+
+	if !a.SessionHasPermissionToChannel(*a.Session(), channelId, model.PERMISSION_UPLOAD_FILE) {
+		return nil, model.NewAppError("UploadFiles", "api.file.upload_file.user_not_authorize_upload.app_error",
+			nil, "", http.StatusBadRequest)
```
Comment (agarciamontoro) [inline app/file.go]: I think I'd copy here what you did above, adding the channel ID, as in `map[string]interface{}{"channelId": channelId}`, so we can use it later in the error.

File: i18n/es.json [hunk 8]
```diff
@@ -1192,6 +1192,14 @@
...
+  },
+  {
+    "id": "api.file.upload_file.user_not_authorize_upload.app_error",
+    "translation": "No se pueden cargar archivos. Se requiere permiso de carga."
```
Comment (agarciamontoro) [inline i18n/es.json]: And the same here: `No tienes los permisos necesarios para subir un archivo al canal con identificador {{.channelId}}.`

File: app/file.go [hunk 9]
```diff
@@ -441,6 +441,17 @@ func (a *App) UploadFile(data []byte, channelId string, filename string) (*model
...
 	}
 
+	_, err := a.GetChannel(channelId)
+	if err != nil {
+		return nil, model.NewAppError("UploadFiles", "api.file.upload_file.incorrect_channelId.app_error",
+			map[string]interface{}{"channelId": channelId}, "", http.StatusBadRequest)
+	}
+
+	if !a.SessionHasPermissionToChannel(*a.Session(), channelId, model.PERMISSION_UPLOAD_FILE) {
+		return nil, model.NewAppError("UploadFiles", "api.file.upload_file.user_not_authorize_upload.app_error",
```
Comment (agarciamontoro) [inline app/file.go]: Minor change in the function name:
```suggestion
		return nil, model.NewAppError("UploadFile", "api.file.upload_file.incorrect_channelId.app_error",
			map[string]interface{}{"channelId": channelId}, "", http.StatusBadRequest)
	}

	if !a.SessionHasPermissionToChannel(*a.Session(), channelId, model.PERMISSION_UPLOAD_FILE) {
		return nil, model.NewAppError("UploadFile", "api.file.upload_file.user_not_authorize_upload.app_error",
```

**Comment by levb** [comment]
Build translation errors need to be fixed? @waqasraz

**Comment by waqasraz** [comment]
> Build translation errors need to be fixed? @waqasraz

how can I do that? I am new here.

**Comment by levb** [comment]
See the above comment from @agarciamontoro

**Comment by agarciamontoro** [comment]
Thank you for addressing my comments, @waqasraz! I've done a bit of research on the translation string issues and you need to run `make i18n-extract`, which generates the English strings from the code (you can add the English strings to the generated fields in the json file, but any other language will be ignored, AFAIK)  and push the changes. This is documented only in the [WebApp wokflow](https://developers.mattermost.com/contribute/webapp/developer-workflow/), but I'll definitely add it to the [Server workflow](https://developers.mattermost.com/contribute/server/developer-workflow/), too. Sorry for the lack of documentation there!

Apart from that, looks like there is another test failing; namely, the [TestCreatePost/with_file_uploaded_by_nouser test](https://github.com/mattermost/mattermost-server/blob/master/api4/post_test.go#L97-L113). Can you take a look at that, too?

**Comment by waqasraz** [comment]
> Apart from that, looks like there is another test failing; namely, the [TestCreatePost/with_file_uploaded_by_nouser test](https://github.com/mattermost/mattermost-server/blob/master/api4/post_test.go#L97-L113). Can you take a look at that, too?

Any Idea what it could mean "with file uploaded by nouser" nouser is allowed file upload? I can create a session and assign file upload permission to the user.

**Comment by agarciamontoro** [comment]
Thank you for keeping up the great work, @waqasraz!

> Any Idea what it could mean "with file uploaded by nouser" nouser is allowed file upload? I can create a session and assign file upload permission to the user.

Yeah, maybe this test makes no sense anymore, as we do need a user to upload a file now! What do you think, @levb?

> For the translation, I did exactly what was told, but still failing. A lot of translation got removed for some reason.

This is probably because your branch was out of sync with master. I triggered an automatic merge of master to see if this makes things better. Let's see if the CI does not complain anymore!

**Comment by levb** [comment]
> > Any Idea what it could mean "with file uploaded by nouser" nouser is allowed file upload? I can create a session and assign file upload permission to the user.
> 
> Yeah, maybe this test makes no sense anymore, as we do need a user to upload a file now! What do you think, @levb?

I don't really understand the file upload permissions model - I only worked on refactoring the code for performance. We'd need to dig deeper to see if anything would break because of this requirement, I don't have any knowledge.

File: app/file_test.go [hunk 10]
```diff
@@ -87,14 +87,21 @@ func TestDoUploadFile(t *testing.T) {
...
 	filename := "test"
 	data := []byte("abcd")
 
-	info1, err := th.App.UploadFile(data, channelId, filename)
+	info1, err := th.App.UploadFile(data, "wrong", filename)
+	require.Error(t, err, "Wrong Channel ID.")
+	require.Nil(t, info1, "Channel ID does not exist.")
+
+	info1, err = th.App.UploadFile(data, "", filename)
+	require.Nil(t, err, "for legacy reasons it should accepts the empty channel id")
```
Comment (agarciamontoro) [inline app/file_test.go]: Nit: no need to explicitly say "for legacy reasons", we can reword this as
```suggestion
	require.Nil(t, err, "empty channel IDs should be valid")
```

File: i18n/en.json [hunk 11]
```diff
@@ -1352,6 +1352,10 @@
     "id": "api.file.test_connection.s3.connection.app_error",
     "translation": "Bad connection to S3 or minio."
   },
+  {
+    "id": "api.file.upload_file.incorrect_channelId.app_error",
+    "translation": "Unable to upload file(s). Incorrect ChannelId: {{.channelId}}"
```
Comment (agarciamontoro) [inline i18n/en.json]: Another nit: UploadFile only allows for uploading one file, so we can make this always singular.
```suggestion
    "translation": "Unable to upload the file. Incorrect channel ID: {{.channelId}}"
```

**Comment by agarciamontoro** [comment]
Thank you again, @waqasraz. Let's wait for the QA team to review the changes and then we can merge your contribution! :)

**Comment by agarciamontoro** [comment]
Just one more time, there was an error in master

**Comment by levb** [comment]
@prapti ping

**Comment by teresa-novoa** [comment]
@prapti friendly reminder to check tis PR, thanks!

**Comment by prapti** [comment]
Added to test queue.

**Comment by prapti** [comment]
@agarciamontoro Thank you so much for the test steps in the ticket. That and the description helped a lot. I verified that there is no error for empty channel ID and valid channel ID, and that there is an "Incorrect channel ID" error for invalid channel. Is there anything that needs to be tested on the UI, such as look for any file that gets uploaded on the channel? Or is this out of scope (needs another function call) to test?

**Comment by agarciamontoro** [comment]
@prapti Thank you for testing this! The only thing I can think of is a smoke test uploading a file to some channels, just to make sure we did not break anything in that regard. That would be awesome!

File: app/file.go [hunk 12]
```diff
@@ -439,6 +439,11 @@ func (a *App) UploadFile(data []byte, channelId string, filename string) (*model
 	if appError != nil {
 		return nil, appError
 	}
+	_, err := a.GetChannel(channelId)
```
Comment (lieut-data) [inline app/file.go:442]: Should we run this check at the top of the method? It seems like the file has already been uploaded at this point.

Comment (agarciamontoro) [reply app/file.go:442]: You're right, I completely missed this in my review! @waqasraz would you be open to move this at the top of the method?

**Comment by agarciamontoro** [comment]
Thank you, @waqasraz! Looks like the linter is complaining about some spacing issues in your latest changes.
In general, you can check the styling with `make check-style` to make sure that the linter will be happy. In this specific case, running `gofmt -w -s app/file.go` should fix the spacing issue.

**Comment by agarciamontoro** [comment]
Thank you for formatting the code, @hanzei!

@prapti, I don't know whether you had the time for that smoke testing so we can merge this. Thanks!

**Comment by agarciamontoro** [comment]
Thank you for the review, @prapti!

And thank you again for the contribution and your patience, @waqasraz :tada: I'm looking forward to seeing more contributions from you!

**Comment by justinegeffen** [comment]
I'm following up the Plugin API docs here: https://developers.mattermost.com/extend/plugins/server/reference/#API and in the process of confirming the doc update process.

**Comment by agarciamontoro** [comment]
@justinegeffen, this one should not have any effect in the documentation, as only the implementation changed.
